src/TorchSharp/JIT/CompilationUnit.cs

```diff
@@ -0,0 +1,137 @@
+
+namespace TorchSharp
+{
+    public static partial class torch
+    {
+        public static partial class jit
+        {
+            /// <summary>
+            /// Represents a TorchScript compilation unit, i.e. a Python script file.
+            /// </summary>
+            /// <example>
+            /// var cu = torch.jit.compile(@"
+            ///   def relu_script(a, b):
+            ///     return torch.relu(a + b)
+            /// ");
+            ///
+            /// var y = cu.invoke("relu_script", torch.randn(10));
+            /// </example>
+            /// <remarks>
+            /// Currently, scripts are limited to defining functions. Classes will be ignored.
+            /// </remarks>
+            public class CompilationUnit : IDisposable
+            {
+                internal CompilationUnit(IntPtr handle) 
+                {
+                    this.handle = handle;
+                }
+
+                ~CompilationUnit() => Dispose(false);
+
+                /// <summary>
+                /// Releases the storage.
+                /// </summary>
+                public void Dispose()
+                {
+                    GC.SuppressFinalize(this);
+                    Dispose(true);
+                }
+
+                /// <summary>
+                /// Implements the .NET Dispose pattern.
+                /// </summary>
+                protected virtual void Dispose(bool disposing)
+                {
+                    if (disposing && handle != IntPtr.Zero) {
+                        handle = IntPtr.Zero;
+                    }
+                }
+
+                [DllImport("LibTorchSharp")]
+                private static extern void THSJIT_CompilationUnit_dispose(IntPtr handle);
+
+                internal IntPtr handle;
+
+                [DllImport("LibTorchSharp")]
+                private static extern void THSJIT_CompilationUnit_Invoke(IntPtr module, string name, IntPtr tensors, int length, AllocatePinnedArray allocator, out sbyte typeCode);
+
+                /// <summary>
+                /// Invoke a function from the compilation unit.
+                /// </summary>
+                /// <param name="name">The name of the function.</param>
+                /// <param name="objs">Function arguments.</param>
+                public object invoke(string name, params object[] objs)
+                {
+                    if (String.IsNullOrEmpty(name)) throw new ArgumentNullException("method name");
+
+                    if (!objs.All(o => typeof(Tensor).IsAssignableFrom(o.GetType()))) {
+                        throw new NotImplementedException($"CompilationUnit.{name}() is not yet taking non-tensors as input arguments");
+                    }
+
+                    IntPtr[] ptrArray = null;
+                    sbyte typeCode = 0;
+
+                    using (var parray = new PinnedArray<IntPtr>()) {
+
+                        var tensors = objs.Select(o => (Tensor)o).ToArray();
```

Comment:
do we need to do that? 

is it possible we cast the array items to `Tensot` inside the loop 

```
for (var i = 0; i < tensors.Length; i++) tensorRefs[i] = tensors[i].Handle;
```